VL blog (#4)
mkondratek/git-machete-blog-post: blog.md

## Take a look at your repository from a new perspective with Git Machete Plugin!

Git and IntelliJ-based software are currently one of the most important tools in software development.
The version control system provided with the JetBrains's IDE products
allows performing pretty much all everyday VCS actions directly via the IDE.
Despite the many benefits of the provided IDE and VCS integration,
the state of more extensive (in terms of branch count) repositories presented in Git Log
often gets non-intuitive and difficult to interpret.

> Creating and maintaining small, focused, easy-to-review PRs is a comfortable and desired approach

Reviewer: I'd emphasize:
```suggestion
Creating and maintaining **small, focused, easy-to-review PRs** is a comfortable and desired approach
```

for both the authors and, especially, the reviewers.
We created a plugin that eases the management of 2+ branches in the repository
while also promoting creation of small PRs &mdash;
**a git repository organizer, and a rebase/merge/push/pull automation tool, named Git Machete.**

It allows for looking at the state of the repository from a previously inexperienceable perspective.
It puts the primary focus on branches (rather than commits),
their relationship to each other and also to the remote repository.
Moreover, it gives both diagnostic and restorative functionalities
for the shape of the repository and the mentioned relationships.
It finds an outstanding application in maintaining linear Git history.

### Git Machete Status Graph

![Git Machete Tab with demo repository](gmt_dark.png "Git Machete Tab")

The picture above presents the demo repository `filter-service` with a few branches.
It contains two **root branches** &mdash; `develop` and `master`.
The first of them has two **child branches** &mdash; `allow-ownership-link`

> (with another child branch &mdash; `build-chain`), and `call-ws`.

Reviewer: ```suggestion
(with a child branch of its own &mdash; `build-chain`), and `call-ws`.
```

Each line contains a name of a branch, an optional custom annotation, and relation to the remote branch.

Without going into details, the **green edge** indicates the desired state &mdash;
the tip of the child branch is a direct descendant of the parent branch tip
(i.e. all commits reachable from the parent are also reachable from the child).<br/>
On the other hand, the **red edge** tells us the opposite &mdash;
the tip of the child branch is **not** a direct descendant of the parent branch tip.

In most cases, an edge becomes red after committing some changes to the parent branch (review fixes)
or pulling new commits to the parent from a remote.

### Machete File

![Machete file](machete-file_dark.png "Machete file")

The branches presented within the graph are defined in a very simple configuration file (`.git/machete`),
which can be easily opened from the plugin's toolbar.
It supports both syntax highlighting and completion of local branch names.
The graph is automatically refreshed when `.git/machete` is modified.

Bear in mind that there is no need to edit the machete file manually to set up your branch layout tree.

> Feel free to use `Discover Branch Layout` functionality that will find it out for you automatically.

Reviewer: Pls mention that the file is generated automatically at the first use

Git and IntelliJ-based software are currently one of the most important tools in software development.
The version control system provided with the JetBrains's IDE products
allows performing pretty much all everyday VCS actions directly via the IDE.

> Despite the many benefits of the provided IDE and VCS integration,

Reviewer: :laughing:

**a git repository organizer, and a rebase/merge/push/pull automation tool, named Git Machete.**

It allows for looking at the state of the repository from a previously inexperienceable perspective.
It puts the primary focus on branches (rather than commits),
their relationship to each other and also to the remote repository.

> Moreover, it gives both diagnostic and restorative functionalities

Reviewer: Hmmm this sentence sounds a bit too formal/hazy... let's just skip it completely TBH

The plugin provides a bunch of actions that can help you to manage your repository.

Among the others, there is `Rebase onto Parent`
(that performs interactive rebase of a given branch onto its parent), `Push` and `Pull`.
Let's have a look at the following example...

![Git Machete - Actions](actions-before_dark.png "Git Machete - Actions - Before")

Branch `allow-ownership-link` is **out of sync** (red edge) to its parent &mdash; `develop`.
Furthermore, it is **ahead of** its remote branch (`origin/develop`).
With Git Machete we can quickly and conveniently lead it to a desired state.

First, `Checkout and Rebase Branch Onto Parent`.
This will perform an interactive rebase from IntelliJ.

> Once the rebase is finished the branch is **in sync** (green edge) with its parent.

Reviewer: For clarity:
```suggestion
Once the rebase is finished, the branch is **in sync** (green edge) with its parent.
```
